TCP transport

rneswold opened the pull request:
This branch contains the proposed changes to support TCP connections to DPM. This is slated to be part of v1.0.0 of the package, so it needs to work as well as the ACNET version did in v.0x.

To-do list:

- [x] Run more aggressive scripts against it
- [x] Propose API changes because this is targeted for v1.x and we don't want to make breaking changes anytime soon. No, we're going to run with this API.
- [x] Use `protobufs` for the protocol (see previous concerns)? No, we'll switch over at a later time. The users are mostly protected from the raw protocol messages so updating later on, theoretically, shouldn't show changes to them.
- [x] Point the package to `acsys-proxy.fnal.gov` (092d941).

rneswold committed:
refactor: move documentation.

The property fields had docstrings associated with them. However, the
recommended way to document them is to add it to the class docstring.
This approach also works when properties are generated using the `attr`
package.

rneswold committed:
fix: mention some APIs no longer use ACNET.

rneswold committed:
refactor: remove `available_dpms()` and `find_dpm()`.

In the TCP world, these functions aren't necessary (and can't be
implemented because the service discovery is done by another layer.)

rneswold committed:
refactor: the DPM module now uses the TCP interface.

rneswold committed:
fix: update script to use the v1.x API.

rneswold committed:
add: add v1.x API-aware script which does a setting.

beauremus commented on acsys/dpm/__init__.py:
```diff
@@ -6,109 +6,90 @@
+                                    unmarshal_reply)
 
-_log = logging.getLogger(__name__)
+_log = logging.getLogger('acsys')
```
This, https://docs.python.org/3.6/library/logging.html#logger-objects, recommends using `__name__` for modules to match the package hierarchy. Do you have an idea of why that doesn't work for us?

rneswold replied:
I was being a bit lazy. In my test scripts, I get the `acsys` logger and set the level. I guess the test scripts can grab each `logger` for each sub package and set the level.

rneswold replied:
Fixed in e3725eb.

beauremus commented on acsys/dpm/__init__.py:
```diff
@@ -144,113 +131,206 @@ class ItemStatus(_ItemCommon):
-
-class DPM:
-    def __init__(self, con, node):
+    def __init__(self):
```
Does `DPM` need `super().__init__()`? I see that it's not in the documentation examples, but from everything else, I know it should be used to inherit.

rneswold replied:
I don't specify a base class so I didn't think it was necessary.

beauremus replied:
Is `asyncio.Protocol` not the base class?

rneswold replied:
You're right; I was just looking at the snippet above. To be safe, I'll add the `super.init()` call.

rneswold replied:
Fixed: d6de4b7

beauremus commented on acsys/dpm/__init__.py:
```diff
@@ -287,157 +366,97 @@ def _xlat_reply(self, msg):
+    async def connect(self):
+        loop = asyncio.get_event_loop()
+        con_fut = loop.create_connection(lambda: self,
+                                         host='dce46.fnal.gov',
```
Just a reminder that we need to change this to `acsys-proxy` before merging.

rneswold replied:
I added it to the checklist in the pull request.

beauremus reviewed (changes requested):
This looks good. As you mentioned we should test it some more. I'm flagging this review as "request changes" so that we can modify the end point before merging.

rneswold committed:
fix: use module name for logger instance.

rneswold committed:
fix: initialize the super-class.

rneswold committed:
fix: move the port back to 6802.

Charlie had moved it to 6805 for testing but the final port address is
6802.

rneswold committed:
refactor: optimize `consumeRawInt()`.

Profiling showed this function was called the most. This commit employs
an optimization which made it 18% faster. Instead of looping with the
iterator, we use `islice()` to quickly transfer the data to a buffer and
then use `int.from_bytes()` to convert it.

rneswold committed:
refactor: use `next()` builtin.

Instead of calling `.__next__()`, which looks ugly, use the proper call
of `next()` (which then calls `.__next__()`.)

rneswold committed:
refactor: no need to range check.

No need to range-check the result since `consumeRawInt` only accepts
integers up to 8 bytes in size.

rneswold committed:
fix: add a `port` parameter to `DpmContext`.

The operational proxy is at port 6802 on *acsys-proxy*. This is now the
default configuration. When we want to test a new proxy, it'll run on
another port of the same machine. This parameter allows us to easily
change a script to use the development proxy.

rneswold committed:
optimize: Optimize and use `unmarshal_int16()`.

`unmarshal_int16()` used to call the more-general `consumeRawInt()` and
then would check the range. It now reads the tag field and decodes one
of the two possible ways the decoding could go.

All routines which unmarshal 16-bits integers (field name decoding and
message name decoding) now call this routine.
